### PR TITLE
feat: support typescript configs using `ts-node`

### LINKED ISSUE #318: feat: support config file as typescript
> @AriPerkkio since the codebase is now in TypeScript (so there are type definitions) would you be open to supporting `eslint-remote-tester.config.ts` using `ts-node` similar to how it's done in Webpack & Jest (among others)?
> I'm happy to have a go at implementing if you are (I've done it in a few projects now - it's typically a pretty straightforward & small diff)

_Originally posted by @G-Rath in https://github.com/AriPerkkio/eslint-remote-tester/issues/89#issuecomment-1001107711_

This change is welcome @G-Rath! I've recently seen these typescript config files with `rollup.config.ts` and it improved developer experience a lot. 

I think this is the spot where current `eslint-remote-tester.config.js` is read using `require` statement. 

https://github.com/AriPerkkio/eslint-remote-tester/blob/1653eba42b43192ed6752a98f2780f505b7ef49b/lib/config/config.ts#L59-L62

Some dependent projects are still plain Javascript code bases so it is important to be backwards compatible with those ones. Maybe `ts-node` could be run only when extension is `.ts`? I'm not familiar with implementation of this kind of support. 

I'm happy to help if you run into any issues.

### PR DESCRIPTION
**Pull request by G-Rath**
This should be all that's required to support typescript configs. This handles supporting explicit configs - I'll do a follow-up PR to allow the default config to be picked up if in TS.

@AriPerkkio could you give me some pointers on how you'd like to test this?

Closes #318

### CODE REVIEW

**Comment by AriPerkkio** [comment]
> how you'd like to test this?

For happy-path test I'd like to see new integration test. These tests create new configuration file and run production build using `node-pty`. https://github.com/AriPerkkio/eslint-remote-tester/blob/master/test/integration/integration.test.ts

Currently each integration test creates new configuration file:

https://github.com/AriPerkkio/eslint-remote-tester/blob/1653eba42b43192ed6752a98f2780f505b7ef49b/test/utils.ts#L73

https://github.com/AriPerkkio/eslint-remote-tester/blob/1653eba42b43192ed6752a98f2780f505b7ef49b/test/utils.ts#L37

These files are Javascript CommonJS modules: 
https://github.com/AriPerkkio/eslint-remote-tester/blob/1653eba42b43192ed6752a98f2780f505b7ef49b/test/utils.ts#L61

I think we could pass a flag from `integration.test.ts` via `function runProductionBuild` which would indicate a TypeScript configuration file should be created. 

```diff
+ if(someOptionIndicatingTS) {
+     fs.writeFileSync(`${name}.ts`, `export default ${configText}`, 'utf8');
+ } else { 
-     fs.writeFileSync(name, `module.exports=${configText}`, 'utf8');
+     fs.writeFileSync(`${name}.js`, `module.exports=${configText}`, 'utf8');
+ }
```

For error case testing I think the `loadConfig` function could be exported and tested in `test/unit/config.test.ts`.

File: test/integration/integration.test.ts [hunk 1]
```diff
@@ -1135,3 +1135,320 @@ test('calls eslintrc function with repository and its location', async () => {
         "
     `);
 });
+
+describe('some tests with typescript configs', () => {
```
Comment (AriPerkkio) [inline test/integration/integration.test.ts]: Looks good, but let's replace these with a single test which validates we can actually pass Typescript into the configuration file, and see it executing:

```ts
test('loads typescript configuration file', async () => {
    const { output } = await runProductionBuild(
        {
            onComplete: `function onComplete() {
                console.log('[TEST-ON-COMPLETE-START]');

                type SomeType = 'a' | 'b' | 'c';
                const value: SomeType = 'b';
                console.log('Value of typed const', value);

                console.log('config file:');
                const index = process.argv.findIndex(a => a === '--config');
                const config = process.argv[index + 1];
                console.log(require('fs').readFileSync(config, 'utf8'));

                console.log('[TEST-ON-COMPLETE-END]');
            }` as unknown as Config['onComplete'],
        },
        undefined,
        'ts'
    );

    const [onCompleteCall] = output
        .join('\n')
        .match(/\[TEST-ON-COMPLETE-START\]([\s|\S]*)\[TEST-ON-COMPLETE-END\]/)!;

    expect(onCompleteCall).toMatchInlineSnapshot();
});
```

If typescript compiling is removed from code base we will see following error:
```ts
    expect(output).toMatchInlineSnapshot(`
        Array [
          "<removed>/test/integration/integration.config-0.js",
          "                type SomeType = 'a' | 'b' | 'c';",
          "                     ^^^^^^^^",
          "SyntaxError: Unexpected identifier",
          "    at Object.compileFunction (node:vm:352:18)",
          "    at wrapSafe (node:internal/modules/cjs/loader:1031:15)",
          "    at Module._compile (node:internal/modules/cjs/loader:1065:27)",
          "    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)",
          "    at Module.load (node:internal/modules/cjs/loader:981:32)",
          "    at Function.Module._load (node:internal/modules/cjs/loader:822:12)",
          "    at Module.require (node:internal/modules/cjs/loader:1005:19)",
          "    at require (node:internal/modules/cjs/helpers:94:18)",
          "    at loadConfig (<removed>/dist/config/load.js)",
          "    at Object.<anonymous> (<removed>/dist/config/config.js)",
        ]
    `);
```

**Comment by AriPerkkio** [comment]
Thanks, looks good! I'll do some more manual testing later before merging this. Also some documentation updates to `README.md` are required.

I just realized we'll need some changes to `eslint-remote-tester-run-action` as well. Currently it holds an internal `eslint-remote-tester.config.js` which is extended by users configuration. I think there it would be enough to check whether user passed a `.ts` config in action parameters, and create internal configuration as typescript. I'll need to think this through. 

https://github.com/AriPerkkio/eslint-remote-tester-run-action/blob/223700ccc8394c4e17d54e0d9d5d43dba8fed06c/src/run-tester.ts#L29-L30

File: test/integration/integration.test.ts [hunk 2]
```diff
@@ -1135,3 +1136,66 @@ test('calls eslintrc function with repository and its location', async () => {
...
     `);
 });
+
+test('loads typescript configuration file', async () => {
+    const { output } = await runProductionBuild(
+        {
+            CI: true,
+            rulesUnderTesting: [],
+            eslintrc: { root: true },
+            onComplete: `function onComplete() {
```
Comment (G-Rath) [inline test/integration/integration.test.ts:1146]: I didn't realise that this'd work - in that case, we should remove the `// @ts-nocheck` [here](https://github.com/AriPerkkio/eslint-remote-tester/pull/320/files#diff-1a258b987746216dc81c746069ca5944c0f1ec848aa095a8a9f109a976dc2343R64) as we should be able to properly type this function now.

**Review by AriPerkkio — APPROVED** [review]
Merging these functional changes as is, in order to keep PR small. Let's continue documentation and configuration type exporting at #321.

> I'll do a follow-up PR to allow the default config to be picked up if in TS.

This can also be started now if you're up for it @G-Rath. Thanks!

**Comment by AriPerkkio** [comment]
@G-Rath as part of ESLint v9 migration (https://github.com/AriPerkkio/eslint-remote-tester/pull/519) I'm converting the configuration loading to ESM. Any ideas how to register `ts-node` for ESM?

**Comment by AriPerkkio** [comment]
I think I'll switch it to https://github.com/antfu/importx instead.

**Comment by G-Rath** [comment]
@AriPerkkio it should be doable though I've not done it myself - `importx` looks cool, though it seems like it directly depends on all the loaders it supports, which doesn't seem ideal (they should be peer dependencies instead)

**Comment by AriPerkkio** [comment]
Good point, install size of `importx` is quite much. 

Looks like ESLint itself has chosen `jiti`. https://github.com/eslint/eslint/pull/18134/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R172-R179
I'll switch `ts-node` to `jiti` so users can share the peer dependency on both projects.

**Pull request merged**